File: src/api/shell/guide_chat.py
"""Guide chat routes — sessions + SSE message stream."""

from __future__ import annotations

import json
from typing import Iterator

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse

from src.api.deps import get_guide_chat_service
from src.api.schemas.guide_chat import (
    GuideMessageListOut,
    GuideMessageOut,
    GuideSendMessageIn,
    GuideSessionListOut,
    GuideSessionOut,
)
from src.app.constants import DEFAULT_MESSAGE_LIMIT, MAX_PAGE_LIMIT
from src.app.services.guide_chat_service import GuideChatService

router = APIRouter(tags=["Guide Chat"])


def _sse_pack(event: str, data: dict) -> str:
    payload = json.dumps(data, ensure_ascii=False)
    return f"event: {event}\ndata: {payload}\n\n"


@router.get("/workspaces/{workspace_id}/guide-sessions", response_model=GuideSessionListOut)
def list_guide_sessions(
    workspace_id: str,
    svc: GuideChatService = Depends(get_guide_chat_service),
) -> GuideSessionListOut:
    items = svc.list_sessions(workspace_id)
    return GuideSessionListOut(
        items=[
            GuideSessionOut(
                id=s.id,
                title=s.title,
                created_at=s.created_at,
                updated_at=s.updated_at,
            )
            for s in items
        ]
    )


@router.post("/workspaces/{workspace_id}/guide-sessions", response_model=GuideSessionOut)
def create_guide_session(
    workspace_id: str,
    svc: GuideChatService = Depends(get_guide_chat_service),
) -> GuideSessionOut:
    s = svc.create_session(workspace_id)
    return GuideSessionOut(
        id=s.id,
        title=s.title,
        created_at=s.created_at,
        updated_at=s.updated_at,
    )


@router.delete("/workspaces/{workspace_id}/guide-sessions/{session_id}", status_code=204)
def delete_guide_session(
    workspace_id: str,
    session_id: str,
    svc: GuideChatService = Depends(get_guide_chat_service),
) -> None:
    svc.delete_session(workspace_id, session_id)


@router.get(
    "/workspaces/{workspace_id}/guide-sessions/{session_id}/messages",
    response_model=GuideMessageListOut,
)
def list_guide_messages(
    workspace_id: str,
    session_id: str,
    limit: int = Query(default=DEFAULT_MESSAGE_LIMIT, ge=1, le=MAX_PAGE_LIMIT),
    offset: int = Query(default=0, ge=0),
    svc: GuideChatService = Depends(get_guide_chat_service),
) -> GuideMessageListOut:
    items, total = svc.list_messages(
        workspace_id, session_id, limit=limit, offset=offset
    )
    return GuideMessageListOut(
        items=[
            GuideMessageOut(
                id=m.id,
                role=m.role,
                text=m.text,
                at=m.at,
                steps=m.steps,
            )
            for m in items
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.post("/workspaces/{workspace_id}/guide-sessions/{session_id}/messages/stream")
def stream_guide_message(
    workspace_id: str,
    session_id: str,
    body: GuideSendMessageIn,
    svc: GuideChatService = Depends(get_guide_chat_service),
) -> StreamingResponse:
    def event_iter() -> Iterator[str]:
        for evt in svc.stream_turn(workspace_id, session_id, body.text):
            yield _sse_pack(str(evt.get("event") or "message"), evt.get("data") or {})

    return StreamingResponse(
        event_iter(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: frontend/src/api/guideChat.ts
/** Guide chat (sessions + SSE) API client. */

import { apiFetch, parseError } from "./client";

export type GuideChatRole = "user" | "assistant";

export interface GuideSession {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
}

export interface GuideStep {
  id: string;
  phase: string;
  title: string;
  detail: string;
}

export interface GuideMessage {
  id: string;
  role: GuideChatRole;
  text: string;
  at: string;
  steps?: GuideStep[] | null;
}

export interface GuideMessageListOut {
  items: GuideMessage[];
  total: number;
  limit: number;
  offset: number;
}

export type GuideStreamHandlers = {
  onStep?: (step: GuideStep) => void;
  onClarify?: (text: string) => void;
  onFinal?: (message: GuideMessage) => void;
  onError?: (message: string) => void;
  onDone?: () => void;
};

export function listGuideSessions(workspaceId: string): Promise<{ items: GuideSession[] }> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/guide-sessions`,
  );
}

export function createGuideSession(workspaceId: string): Promise<GuideSession> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/guide-sessions`,
    { method: "POST", body: "{}" },
  );
}

export function deleteGuideSession(workspaceId: string, sessionId: string): Promise<void> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/guide-sessions/${encodeURIComponent(sessionId)}`,
    { method: "DELETE" },
  );
}

export function listGuideMessages(
  workspaceId: string,
  sessionId: string,
  opts?: { limit?: number; offset?: number },
): Promise<GuideMessageListOut> {
  const sp = new URLSearchParams();
  if (opts?.limit != null) sp.set("limit", String(opts.limit));
  if (opts?.offset != null) sp.set("offset", String(opts.offset));
  const q = sp.toString() ? `?${sp}` : "";
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/guide-sessions/${encodeURIComponent(sessionId)}/messages${q}`,
  );
}

function parseSseChunk(
  buffer: string,
  onEvent: (event: string, data: unknown) => void,
): string {
  const parts = buffer.split("\n\n");
  const rest = parts.pop() ?? "";
  for (const block of parts) {
    if (!block.trim()) continue;
    let event = "message";
    const dataLines: string[] = [];
    for (const line of block.split("\n")) {
      if (line.startsWith("event:")) event = line.slice(6).trim();
      else if (line.startsWith("data:")) dataLines.push(line.slice(5).trim());
    }
    const raw = dataLines.join("\n");
    if (!raw) continue;
    try {
      onEvent(event, JSON.parse(raw));
    } catch {
      onEvent(event, { message: raw });
    }
  }
  return rest;
}

export async function streamGuideMessage(
  workspaceId: string,
  sessionId: string,
  text: string,
  handlers: GuideStreamHandlers,
): Promise<void> {
  const res = await fetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/guide-sessions/${encodeURIComponent(sessionId)}/messages/stream`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "text/event-stream" },
      body: JSON.stringify({ text }),
    },
  );
  if (!res.ok) throw new Error(await parseError(res));
  if (!res.body) throw new Error("No response body for SSE stream");

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  let finished = false;
  const finish = () => {
    if (finished) return;
    finished = true;
    handlers.onDone?.();
  };

  const dispatch = (event: string, data: unknown) => {
    const obj = (data && typeof data === "object" ? data : {}) as Record<string, unknown>;
    if (event === "step") {
      handlers.onStep?.({
        id: String(obj.id ?? ""),
        phase: String(obj.phase ?? ""),
        title: String(obj.title ?? ""),
        detail: String(obj.detail ?? ""),
      });
    } else if (event === "clarify") {
      handlers.onClarify?.(String(obj.text ?? ""));
    } else if (event === "final") {
      const msg = obj.message as GuideMessage | undefined;
      if (msg) handlers.onFinal?.(msg);
    } else if (event === "error") {
      handlers.onError?.(String(obj.message ?? "خطای ناشناخته"));
    } else if (event === "done") {
      finish();
    }
  };

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    buffer = parseSseChunk(buffer, dispatch);
  }
  if (buffer.trim()) {
    buffer = parseSseChunk(buffer + "\n\n", dispatch);
  }
  finish();
}
